### PR TITLE
[ENH] Raise more meaningful configuration warnings.

### PR DESCRIPTION
**Pull request by richard-willdooit**

### CODE REVIEW

**Comment by richard-willdooit** [comment]
@PCatinean Hope you can find this feature valuable.  It should work without requiring a module upgrade, but does have a new field...

File: product_configurator/models/product.py [hunk 1]
```diff
@@ -369,19 +369,25 @@ def value_available(self, attr_val_id, value_ids):
...
-                    return False
-            else:
-                if set(domain[2]) & set(value_ids):
-                    return False
+        for config_line in config_lines:
+            domains = config_line.mapped('domain_id').compute_domain()
+            for domain in domains:
+                if domain[1] == 'in':
+                    if not set(domain[2]) & set(value_ids):
+                        # Don't break if the module has not been upgraded...
```
Comment (PCatinean) [inline product_configurator/models/product.py:378]: Could you add a TODO here to remove the check for field existence in future versions?

File: product_configurator/models/product.py [hunk 2]
```diff
@@ -369,19 +369,25 @@ def value_available(self, attr_val_id, value_ids):
...
+                        if do_raise and 'rule_description' in config_line._fields and config_line.rule_description:
+                            raise ValidationError(config_line.rule_description)
+                        return False
+                else:
+                    if set(domain[2]) & set(value_ids):
+                        # Don't break if the module has not been upgraded...
```
Comment (PCatinean) [inline product_configurator/models/product.py:384]: Here also

File: product_configurator/models/product.py [hunk 3]
```diff
@@ -407,21 +413,31 @@ def validate_configuration(self, value_ids, custom_vals=None, final=True):
...
+                if do_raise:
+                    attr_val = self.env['product.attribute.value'].browse(val)
+                    raise ValidationError(_("%s can not be selected for %s") % (attr_val.name, attr_val.attribute_id.name))
                 return False
 
         # Check if custom values are allowed
         custom_attr_ids = self.attribute_line_ids.filtered(
             'custom').mapped('attribute_id').ids
 
         if not set(custom_vals.keys()) <= set(custom_attr_ids):
```
Comment (PCatinean) [inline product_configurator/models/product.py]: Here maybe we can do invalid_custom_vals = set1 - set2

And if this sets has any values we can pass them in the ValdationError with "Attributes (%s) cannot hold custom values"

What do you think?

**Comment by PCatinean** [comment]
@richard-willdooit the domain parsing part will need a pretty big overhaul (it's the part of the entire project that i'm least proud of).

But otherwise raising meaningful errors was always on the TODO list so thanks for taking care of it.

File: product_configurator/models/product.py (hunk 3 above)
Comment (richard-willdooit) [reply product_configurator/models/product.py]: I have applied a change - but my data and setup cannot raise this type of error, so I have not actually tested and proved it works.  It doesn't fail for my data!

**Comment by PCatinean** [comment]
@richard-willdooit thanks for the updates.

To prevent regressions and maintain quality as much as possible I've added automated tests.

A few commits ago I made everything pass and it seems something is failing on this build.

Could you take a look and see why it breaks?

I want to try have all builds passing before I merge them and later adding even more tests.

**Comment by richard-willdooit** [comment]
@PCatinean Passing for lint checkers now, but failing on code coverage tests - I am sorry, but this is new to me and I am not sure what to do for this to pass...

**Comment by PCatinean** [comment]
@richard-willdooit it's ok I can take over the PR and try to fix it.

I'm not sure how I can do this to preserve you as the author but I think it should be possible.

Thank you for your efforts! I'll do my best to find time and make review all the PR's and issues

**Comment by PCatinean** [comment]
@richard-willdooit since you introduced new code the total percentage of the tested code has fallen.

Code coverage prevents this and does not pass when a commit provides less test converage than the base.

So basically just adding a test to verify your code does what it is supposed to do (raise error when do_raise is passed and invalid conf is passed) and return False it is not should be enough.

**Comment by PCatinean** [comment]
@richard-willdooit on a second inspection, all you have to do is rebase with the current 10.0 version and it should be fine

**Pull request closed without merging**